File: sentil-jl/examples/synthesis.jl
using Sentil

a = reshape([1.0], 1, 1)
b = reshape([1.0], 1, 1)
model = linear_model(a, b, [1.0], ["x"], 1.0, 3)
spec = formula("always (x > 0)")
bounds = Bounds([-1.0, -1.0, -1.0], [1.0, 1.0, 1.0])

result = synthesize(model, spec; bounds = bounds)
println("input: ", round.(result.input; digits = 4),
        "  robustness: ", result.robustness, "  holds: ", result.holds)

shield = SafetyFilter(bounds)
println("shielded: ", safe_input(shield, [2.0, 0.5, -3.0]))

File: sentil-core/examples/synthesis.rs
//! Synthesize a control-input sequence that satisfies a spec on a linear model.
//!
//! Run with `cargo run --example synthesis`.

use sentil::synthesis::{Backend, Bounds, LinearModel, SynthesisProblem, Synthesizer};
use sentil::Formula;

fn main() -> sentil::Result<()> {
    // x_{t+1} = x_t + u_t over three steps; keep x above zero.
    let model = LinearModel::new(vec![vec![1.0]], vec![vec![1.0]], vec![1.0], ["x"], 1.0, 3)?;
    let spec = Formula::parse("always (x > 0)")?;
    let bounds = Bounds::new(vec![-1.0, -1.0, -1.0], vec![1.0, 1.0, 1.0])?;

    let problem = SynthesisProblem::new(&model, &spec)
        .with_bounds(bounds)
        .with_backend(Backend::Gradient)
        .with_budget(200);
    let result = Synthesizer::solve(&problem)?;
    println!(
        "input {:?}, robustness {:.4}, holds {}",
        result.input, result.robustness, result.holds
    );
    Ok(())
}